lib/publiccloud/azure.pm

```diff
@@ -72,18 +73,45 @@ sub find_img {
     my $container = $self->container;
     my $storage_account = get_var('PUBLIC_CLOUD_STORAGE_ACCOUNT', 'eisleqaopenqa');
     my $key = $self->get_storage_account_keys($storage_account);
-    my $cmd_show = "az storage blob show --account-key $key -o json " .
-      "--container-name '$container' --account-name '$storage_account' --name '$name' " .
-      '--query="{name: name,createTime: properties.creationTime,md5: properties.contentSettings.contentMd5}"';
-    record_info('BLOB INFO', script_output($cmd_show, proceed_on_failure => 1));
-
-    my $json = script_output("az image show --resource-group " . $self->resource_group . " --name $name", 60, proceed_on_failure => 1);
-    record_info('IMG INFO', $json);
-    my $image;
+
+    $json = script_output("az storage blob show --account-key $key -o json " .
+          "--container-name '$container' --account-name '$storage_account' --name '$name' " .
+          '--query="{name: name,createTime: properties.creationTime,md5: properties.contentSettings.contentMd5}"',
+        proceed_on_failure => 1);
+    record_info('BLOB INFO', $json);
     eval {
-        $image = decode_azure_json($json)->{name};
+        $md5 = decode_azure_json($json)->{md5};
     };
-    record_info('IMG NOT-FOUND', "Cannot find image $name. Need to upload it.\n$@") if ($@);
+    if ($@) {
```

Comment:
Is this correct? Isn't `if ($@)` being executed, if in the previous `eval` block the json could be parsed?

Reply:
See https://perldoc.perl.org/functions/eval
> If there is a syntax error or runtime error, or a [die](https://perldoc.perl.org/functions/die) statement is executed, eval returns [undef](https://perldoc.perl.org/functions/undef) in scalar context, or an empty list in list context, and [$@](https://perldoc.perl.org/perlvar#%24%40) is set to the error message. (Prior to 5.16, a bug caused [undef](https://perldoc.perl.org/functions/undef) to be returned in list context for syntax errors, but not for runtime errors.) If there was no error, [$@](https://perldoc.perl.org/perlvar#%24%40) is set to the empty string. A control flow operator like [last](https://perldoc.perl.org/functions/last) or [goto](https://perldoc.perl.org/functions/goto) can bypass the setting of [$@](https://perldoc.perl.org/perlvar#%24%40). Beware that using eval neither silences Perl from printing warnings to STDERR, nor does it stuff the text of warning messages into [$@](https://perldoc.perl.org/perlvar#%24%40). To do either of those, you have to use the [$SIG{__WARN__}](https://perldoc.perl.org/perlvar#%25SIG) facility, or turn off warnings inside the BLOCK or EXPR using no warnings 'all'. See [warn](https://perldoc.perl.org/functions/warn), [perlvar](https://perldoc.perl.org/perlvar), and [warnings](https://perldoc.perl.org/warnings).

Reply:
Ah, that's the error message, not the return code. Yep, then it works.